altair/utils/data.py

```diff
@@ -2,63 +2,84 @@
 import os
 import random
 import hashlib
+import sys
 import warnings
+from typing import Union, MutableMapping, Optional, Dict, Sequence, TYPE_CHECKING, List
+from types import ModuleType
 
 import pandas as pd
 from toolz import curried
-from typing import Callable
+from typing import TypeVar
 
 from .core import sanitize_dataframe, sanitize_arrow_table
 from .core import sanitize_geo_interface
 from .deprecation import AltairDeprecationWarning
 from .plugin_registry import PluginRegistry
 
 
+if sys.version_info >= (3, 8):
+    from typing import Protocol, TypedDict, Literal
+else:
+    from typing_extensions import Protocol, TypedDict, Literal
+
+
+if TYPE_CHECKING:
+    import pyarrow.lib
+
+
+class _SupportsGeoInterface(Protocol):
+    __geo_interface__: MutableMapping
+
+
+class _SupportsDataframe(Protocol):
+    def __dataframe__(self, *args, **kwargs):
+        ...
+
+
+_DataType = Union[dict, pd.DataFrame, _SupportsGeoInterface, _SupportsDataframe]
+_TDataType = TypeVar("_TDataType", bound=_DataType)
+
+_VegaLiteDataDict = Dict[str, Union[str, dict, List[dict]]]
+_ToValuesReturnType = Dict[str, Union[dict, List[dict]]]
```

Comment:
I'm not sure the best way to do this, but should the type encode that this dictionary only has a top-level "values" key?

Reply:
Yeah I agree it would be nice and we could use a `TypedDict`. Unfortunately, these lines in the function allow a dictionary to have more then just `values` as a top-level key in case `data` is a dictionary:

```python
    elif isinstance(data, dict):
        if "values" not in data:
            raise KeyError("values expected in data dict, but not present.")
        return data
```

I think there is currently no way to use a `TypedDict` with some fixed keys and then allow for arbitrary extra keys, see [this discussion over in the mypy repo](https://github.com/python/mypy/issues/4617).